### PR TITLE
standalone javascript console

### PR DESCRIPTION
**Pull request by debris**
- new executable `ethconsole` which connect to remote jsonrpc http server and provides javascript console

### CODE REVIEW

File: eth/main.cpp [hunk 1]
```diff
@@ -1742,12 +1742,24 @@ int main(int argc, char** argv)
 		if (useConsole)
 		{
 #if ETH_JSCONSOLE
-			JSConsole console(web3, make_shared<SimpleAccountHolder>([&](){return web3.ethereum();}, getAccountPassword, keyManager));
+			JSLocalConsole console;
+
+			jsonrpcServer = shared_ptr<dev::WebThreeStubServer>(new dev::WebThreeStubServer(*jsonrpcConnector.get(), web3, make_shared<SimpleAccountHolder>([&](){ return web3.ethereum(); }, getAccountPassword, keyManager), vector<KeyPair>(), keyManager, *gasPricer));
```
Comment (LefterisJP) [inline eth/main.cpp]: Do split the arguments per line for visibility

**Comment by LefterisJP** [comment]
Question: Will this only be a standalone? Or can we also add it inside the CLI client itself and replace the current "interactive" mode?

**Comment by debris** [comment]
> Question: Will this only be a standalone? Or can we also add it inside the CLI client itself and replace the current "interactive" mode?

Currently there are two ways of using javascript console.

1st way - similar to interactive mode
```bash
./eth --console
```

2nd way
```bash
# start json-rpc http server
./eth -j

# connect console to json-rpc server
./ethconsole http://127.0.0.1:8545/
```

File: ethconsole/main.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+using namespace dev;
+using namespace dev::eth;
+
+int main(int argc, char** argv)
+{
+	if (argc != 2)
+	{
+		cout << "You must provide remote url\n";
+		cout << "eg:\n";
+		cout << "./ethconsole http://localhost:8545\n";
```
Comment (gavofyork) [inline ethconsole/main.cpp]: should default to exactly that.

File: libjsconsole/CURLRequest.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <stdio.h>
+#include <sstream>
+#include <unordered_map>
+#include <curl/curl.h>
+
+class CURLRequest
+{
+public:
+	CURLRequest(): m_curl(curl_easy_init()) {};
```
Comment (gavofyork) [inline libjsconsole/CURLRequest.h]: trailing `;`

File: libjsconsole/CURLRequest.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	std::tuple<long, std::string> post();
+
+private:
+	std::string m_url;
+	std::string m_body;
+
+	CURL* m_curl;
+	std::stringstream m_resultBuffer;
+
+	void commonCURLPreparation();
```
Comment (gavofyork) [inline libjsconsole/CURLRequest.h]: trailing `;`

File: libjsconsole/CURLRequest.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	~CURLRequest()
+	{
+		if (m_curl)
+			curl_easy_cleanup(m_curl);
+	}
+
+	void setUrl(std::string _url) { m_url = _url; }
+	void setBody(std::string _body) { m_body = _body; }
+
+	std::tuple<long, std::string> post();
```
Comment (gavofyork) [inline libjsconsole/CURLRequest.h]: trailing `;`

File: libjsconsole/CURLRequest.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+private:
+	std::string m_url;
+	std::string m_body;
+
+	CURL* m_curl;
+	std::stringstream m_resultBuffer;
+
+	void commonCURLPreparation();
+	std::tuple<long, std::string> commonCURLPerform();
```
Comment (gavofyork) [inline libjsconsole/CURLRequest.h]: trailing `;`

File: libjsconsole/JSConsole.h [hunk 7]
```diff
@@ -22,32 +22,66 @@
...
 
-class AccountHolder;
-
+template<typename Engine, typename Printer>
 class JSConsole
 {
 public:
-	JSConsole(WebThreeDirect& _web3, std::shared_ptr<AccountHolder> const& _accounts);
-	void readExpression() const;
+	JSConsole(): m_engine(Engine()), m_printer(Printer(m_engine)) {};
```
Comment (gavofyork) [inline libjsconsole/JSConsole.h]: trailing `;` and more follow.

File: libjsconsole/JSV8RemoteConnector.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+namespace dev
+{
+namespace eth
+{
+
+class JSV8RemoteConnector : public JSV8RPC
+{
+
+public:
+	JSV8RemoteConnector(JSV8Engine const& _engine, std::string _url): JSV8RPC(_engine), m_url(_url) {}
```
Comment (gavofyork) [inline libjsconsole/JSV8RemoteConnector.h]: trailing `;`s

File: libjsconsole/JSRemoteConsole.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+namespace dev
+{
+namespace eth
+{
+
+class JSRemoteConsole: public JSConsole<JSV8Engine, JSV8Printer>
+{
+
+public:
+	JSRemoteConsole(std::string _url): m_connector(m_engine, _url) {};
```
Comment (gavofyork) [inline libjsconsole/JSRemoteConsole.h]: trailing `;`s

File: libjsconsole/JSLocalConsole.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace dev
+{
+namespace eth
+{
+
+class JSLocalConsole: public JSConsole<JSV8Engine, JSV8Printer>
+{
+public:
+	JSLocalConsole();
```
Comment (gavofyork) [inline libjsconsole/JSLocalConsole.h]: trailing `;`s.

**Comment by gavofyork** [comment]
loads of trailing `;`s.

**Comment by debris** [comment]
Bad habit :/ Fixed.

File: ethconsole/main.cpp (hunk 2 above)
Comment (gavofyork) [inline ethconsole/main.cpp]: should still default to exactly that.

**Comment by gavofyork** [comment]
no longer defaulting to the sensible server/port.

**Pull request merged**